File: 322.cpp
class Solution {
public:

    int possMin(int a, int b) {
        if (a == -1) return b;
        if (b == -1 && a != -1) return a + 1;
        return min<int>(a + 1, b);
    }
    
    int coinChange(vector<int>& coins, int amount) {
        if (amount == 0) return 0;
        
        int a[amount + 1][coins.size()];
        int x, y, m;
        
        for (int i = 0; i < coins.size(); i++) {
            a[0][i] = 0;
        }
        
        for (int i = 1; i <= amount; i++) {
            for (int j = 0; j < coins.size(); j++) {
                x = (i - coins[j] < 0 || j < 0)? -1 : a[i - coins[j]][j];
                y = (i < 0 || j - 1 < 0)? -1 : a[i][j - 1];
                a[i][j] = possMin(x, y);
            }
        }
        
        m = -1;
        
        for (int i = 0; i < coins.size(); i++) {
            if (a[amount][i] != -1 && (a[amount][i] < m || m == -1)) {
                m = a[amount][i];
            }
        }
        return m;
    }
};

File: 322.py
class Solution:
    def coinChange(self, coins, amount):
        """
        :type coins: List[int]
        :type amount: int
        :rtype: int
        """        
        lis = [[0] * len(coins) for _ in range(amount+1)]
        
        def judge(n , pr):
            if n >= 0 and pr >= 0:
                return lis[n][pr]
            return -1
        
        
        def poss_min(a, b):
            if a == -1:
                return b
            if b == -1 and a != -1:
                return a + 1
            return min(a + 1, b)
        
        for n in range(1, amount + 1):
            for pr in range(len(coins)):
                lis[n][pr] = poss_min(judge(n-coins[pr], pr), judge(n, pr-1))
                
        m = -1
        
        for i in lis[-1]:
            if i != -1 and (i < m or m == -1):
                m = i

        return m